Fixing CMake release build for Win32 target

tcervi opened the pull request:
- Issue #47166 raises the concern about a common build failure when building tensorflowlite_c.dll with cmake and msvc;
- The issue is not related to tensorflow source code, instead it is probably related to a compiler (MSVC) issue when handling /O2 code optimization;
- The proposal is to update CMakeLists.txt so it can recognize a MSVC x86 build and update the compiler optimization flags: from /O2 to /O1;

tcervi committed:
Fixing CMake release build for Win32 target

- Issue #47166 raises the concern about a common build failure when building tensorflowlite_c.dll with cmake and msvc;
- The issue is not related to tensorflow source code, instead it is probably related to a compiler (MSVC) issue when handling /O2 code optimization;
- The proposal is to update CMakeLists.txt so it can recognize a MSVC x86 build and update the compiler optimization flags: from /O2 to /O1;

tcervi commented:
@googlebot I signed it!

abattery commented:
@terryheo could you review this?

terryheo commented on tensorflow/lite/c/CMakeLists.txt:
```diff
@@ -36,6 +36,30 @@ add_subdirectory(
 
 set(CMAKE_CXX_STANDARD 11)
 
+if (CMAKE_SIZEOF_VOID_P EQUAL 4)
```
If this is only needed for Windows build, let's unify this with the condition of line #155

tcervi replied:
Sure thing, did you mean line 45 though?

terryheo replied:
There is a block with `if(CMAKE_SYSTEM_NAME MATCHES "Windows")`
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/CMakeLists.txt#L155
I think this logic should be inside of the block.

tcervi replied:
Sure thing I could also insert it there. Probably inside the block with ` if(MSVC)` since i'm not sure if other compilers would support/require the same flag changes https://github.com/tensorflow/tensorflow/blob/704610e1d21288482bf77923b387f9bb6c119318/tensorflow/lite/CMakeLists.txt#L162
Was just confused because my PR was originally for the Lite C API CMakeLists, but it could also be useful for the Lite main CMakeLists, since the same build issue occurred las time I tried to build it. Lately i've being building only the C API DLL though.

Should we place it on both CMakeLists so?

terryheo replied:
I think we only need to update the main CMakeLists.txt file. Could you verify?
If it doesn't work, let's just wrap it with `if(CMAKE_SYSTEM_NAME MATCHES "Windows")`

tcervi replied:
I could be building it "wrong" but was directly calling the lite/c CMakeLists.txt from the command line, using cmake+msvc and building only the c api libraries. But for sure I can verify the update on the main one and check if it works. Will post it back here soon

terryheo commented on tensorflow/lite/c/CMakeLists.txt:
```diff
@@ -36,6 +36,30 @@ add_subdirectory(
+endif()
+
+if (MSVC AND (NOT 64_BIT))
+  message("Disabling MSVC /O2 optimization")
```
"Disabling MSVC /O2 optimization for Win32"

tcervi replied:
Great!

terryheo commented on tensorflow/lite/c/CMakeLists.txt:
```diff
@@ -36,6 +36,30 @@ add_subdirectory(
+if (MSVC AND (NOT 64_BIT))
+  message("Disabling MSVC /O2 optimization")
+  set(CompFlags
+    CMAKE_CXX_FLAGS_RELEASE
```
Will this work with Debug build?

tcervi replied:
No need to disable /O2 and enable /O1 for debug build, since optimisation is disabled by default (/Od) when building Debug libraries

terryheo reviewed (commented):
thanks for your contribution

tcervi committed:
Applying PR code review

tcervi committed:
Adding Win32 build fix to lite main CMakeLists.txt

tcervi committed:
Wrapping Win32 build fix with CAMKE_SYSTEM_NAME if

tcervi committed:
Enhancing code indentation

tcervi commented:
Hi @terryheo 

I'm sorry for my delay but was no longer working on that project so I had to bring up a new Windows VM to check the changes from my side. Apparently it is all good, the build stills working, not sure why the Community CI Pipeline failed for the last commits though, could not understand the issue from Jenkins logs. Could you help me please?

Like you requested, I applied the changes on the main CMakeLists (lite/CMakeLists) and it fixed the C++ tensorflow-lite library build, as far as I can tell. I still confused about the C API though, I was usually only build this library (lite/c/CMakeLists) and that's why I changed the c/CMakeLists file in the begging.  I added the SYSTEM_NAME check as peer your suggestion, but still thinking that it should be on both CMakeLists, since people can need to build one or the other. 

Any thoughts on that?

tcervi committed:
Merge branch 'master' into tflite-c-win32-cmake

tcervi commented:
Updating my PR with master code fixed the CI build issue.
@terryheo could you please review it when you have some time?

Cheers

terryheo reviewed (approved):
LGTM

tcervi committed:
Merge branch 'master' into tflite-c-win32-cmake
